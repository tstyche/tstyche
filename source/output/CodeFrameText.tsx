import type { TestTree, TestTreeNode } from "#collect";
import { DiagnosticCategory, type DiagnosticOrigin, type SourceFile } from "#diagnostic";
import { Path } from "#path";
import { Color, Line, type ScribblerJsx, Text } from "#scribbler";
import { SourceService } from "#source";
import type { CodeFrameOptions } from "./types.js";

interface BreadcrumbsTextProps {
  ancestor: TestTreeNode | TestTree;
}

function BreadcrumbsText({ ancestor }: BreadcrumbsTextProps) {
  const text: Array<string> = [];

  while ("name" in ancestor) {
    if (ancestor.name !== "") {
      text.push(ancestor.name);
    }

    ancestor = ancestor.parent;
  }

  text.push("");

  return <Text color={Color.Gray}>{text.reverse().join(" ‚ù≠ ")}</Text>;
}

interface CodeLineTextProps {
  gutterWidth: number;
  lineNumber: number;
  lineNumberColor?: Color;
  lineText: string;
}

function CodeLineText({ gutterWidth, lineNumber, lineNumberColor = Color.Gray, lineText }: CodeLineTextProps) {
  return (
    <Line>
      <Text color={lineNumberColor}>{lineNumber.toString().padStart(gutterWidth)}</Text>
      <Text color={Color.Gray}>{" | "}</Text>
      {lineText}
    </Line>
  );
}

interface SquiggleLineTextProps {
  gutterWidth: number;
  indentWidth?: number;
  squiggleColor: Color;
  squiggleWidth: number;
}

function SquiggleLineText({ gutterWidth, indentWidth = 0, squiggleColor, squiggleWidth }: SquiggleLineTextProps) {
  return (
    <Line>
      {" ".repeat(gutterWidth)}
      <Text color={Color.Gray}>{" | "}</Text>
      {" ".repeat(indentWidth)}
      <Text color={squiggleColor}>{"~".repeat(squiggleWidth === 0 ? 1 : squiggleWidth)}</Text>
    </Line>
  );
}

interface CodeFrameTextProps {
  diagnosticCategory: DiagnosticCategory;
  diagnosticOrigin: DiagnosticOrigin;
  options?: CodeFrameOptions | undefined;
}

export function CodeFrameText({ diagnosticCategory, diagnosticOrigin, options }: CodeFrameTextProps) {
  const linesAbove = options?.linesAbove ?? 2;
  const linesBelow = options?.linesBelow ?? 3;
  const showBreadcrumbs = options?.showBreadcrumbs ?? true;

  const sourceFile = SourceService.get(diagnosticOrigin.sourceFile);

  const lineMap = sourceFile.getLineStarts();

  const { character: firstMarkedLineCharacter, line: firstMarkedLine } = sourceFile.getLineAndCharacterOfPosition(
    diagnosticOrigin.start,
  );
  const { character: lastMarkedLineCharacter, line: lastMarkedLine } = sourceFile.getLineAndCharacterOfPosition(
    diagnosticOrigin.end,
  );

  const firstLine = Math.max(firstMarkedLine - linesAbove, 0);
  const lastLine = Math.min(lastMarkedLine + linesBelow, lineMap.length - 1);
  const gutterWidth = (lastLine + 1).toString().length + 2;

  let highlightColor: Color;

  switch (diagnosticCategory) {
    case DiagnosticCategory.Error:
      highlightColor = Color.Red;
      break;

    case DiagnosticCategory.Warning:
      highlightColor = Color.Yellow;
      break;
  }

  const codeFrame: Array<ScribblerJsx.Element> = [];

  for (let index = firstLine; index <= lastLine; index++) {
    const lineStart = lineMap[index];
    const lineEnd = index === lineMap.length - 1 ? sourceFile.text.length : lineMap[index + 1];

    const lineText = sourceFile.text.slice(lineStart, lineEnd).trimEnd().replace(/\t/g, " ");

    if (index >= firstMarkedLine && index <= lastMarkedLine) {
      codeFrame.push(
        <CodeLineText
          gutterWidth={gutterWidth}
          lineNumber={index + 1}
          lineNumberColor={highlightColor}
          lineText={lineText}
        />,
      );

      if (index === firstMarkedLine) {
        const squiggleLength =
          index === lastMarkedLine
            ? lastMarkedLineCharacter - firstMarkedLineCharacter
            : lineText.length - firstMarkedLineCharacter;

        codeFrame.push(
          <SquiggleLineText
            gutterWidth={gutterWidth}
            indentWidth={firstMarkedLineCharacter}
            squiggleColor={highlightColor}
            squiggleWidth={squiggleLength}
          />,
        );
      } else if (index === lastMarkedLine) {
        codeFrame.push(
          <SquiggleLineText
            gutterWidth={gutterWidth}
            squiggleColor={highlightColor}
            squiggleWidth={lastMarkedLineCharacter}
          />,
        );
      } else {
        codeFrame.push(
          <SquiggleLineText gutterWidth={gutterWidth} squiggleColor={highlightColor} squiggleWidth={lineText.length} />,
        );
      }
    } else {
      codeFrame.push(<CodeLineText gutterWidth={gutterWidth} lineNumber={index + 1} lineText={lineText} />);
    }
  }

  let breadcrumbs: ScribblerJsx.Element | undefined;

  if (showBreadcrumbs && diagnosticOrigin.assertion != null) {
    breadcrumbs = <BreadcrumbsText ancestor={diagnosticOrigin.assertion.parent} />;
  }

  const location = (
    <Line>
      {" ".repeat(gutterWidth + 2)}
      <Text color={Color.Gray}>{" at "}</Text>
      <Text color={Color.Cyan}>{Path.relative("", diagnosticOrigin.sourceFile.fileName)}</Text>
      <Text color={Color.Gray}>{`:${firstMarkedLine + 1}:${firstMarkedLineCharacter + 1}`}</Text>
      {breadcrumbs}
    </Line>
  );

  return (
    <Text>
      {codeFrame}
      <Line />
      {location}
    </Text>
  );
}
