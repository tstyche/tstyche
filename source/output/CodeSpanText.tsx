import type { TestMember, TestTree } from "#collect";
import { DiagnosticCategory, type DiagnosticOrigin } from "#diagnostic";
import { Path } from "#path";
import { Color, Line, type ScribblerJsx, Text } from "#scribbler";

interface BreadcrumbsTextProps {
  ancestor: TestMember | TestTree;
}

function BreadcrumbsText({ ancestor }: BreadcrumbsTextProps) {
  const text: Array<string> = [];

  while ("name" in ancestor) {
    text.push(ancestor.name);
    ancestor = ancestor.parent;
  }

  text.push("");

  return <Text color={Color.Gray}>{text.reverse().join(" ‚ù≠ ")}</Text>;
}

interface CodeLineTextProps {
  gutterWidth: number;
  lineNumber: number;
  lineNumberColor?: Color;
  lineText: string;
}

function CodeLineText({ gutterWidth, lineNumber, lineNumberColor = Color.Gray, lineText }: CodeLineTextProps) {
  return (
    <Line>
      <Text color={lineNumberColor}>{lineNumber.toString().padStart(gutterWidth)}</Text>
      <Text color={Color.Gray}>{" | "}</Text>
      {lineText}
    </Line>
  );
}

interface SquiggleLineTextProps {
  gutterWidth: number;
  indentWidth?: number;
  squiggleColor: Color;
  squiggleWidth: number;
}

function SquiggleLineText({ gutterWidth, indentWidth = 0, squiggleColor, squiggleWidth }: SquiggleLineTextProps) {
  return (
    <Line>
      {" ".repeat(gutterWidth)}
      <Text color={Color.Gray}>{" | "}</Text>
      {" ".repeat(indentWidth)}
      <Text color={squiggleColor}>{"~".repeat(squiggleWidth === 0 ? 1 : squiggleWidth)}</Text>
    </Line>
  );
}

interface CodeSpanTextProps {
  diagnosticCategory: DiagnosticCategory;
  diagnosticOrigin: DiagnosticOrigin;
}

export function CodeSpanText({ diagnosticCategory, diagnosticOrigin }: CodeSpanTextProps) {
  const lastLineInFile = diagnosticOrigin.sourceFile.getLineAndCharacterOfPosition(
    diagnosticOrigin.sourceFile.text.length,
  ).line;

  const { character: firstMarkedLineCharacter, line: firstMarkedLine } =
    diagnosticOrigin.sourceFile.getLineAndCharacterOfPosition(diagnosticOrigin.start);
  const { character: lastMarkedLineCharacter, line: lastMarkedLine } =
    diagnosticOrigin.sourceFile.getLineAndCharacterOfPosition(diagnosticOrigin.end);

  const firstLine = Math.max(firstMarkedLine - 2, 0);
  const lastLine = Math.min(firstLine + 5, lastLineInFile);
  const gutterWidth = (lastLine + 1).toString().length + 2;

  let highlightColor: Color;

  switch (diagnosticCategory) {
    case DiagnosticCategory.Error: {
      highlightColor = Color.Red;
      break;
    }

    case DiagnosticCategory.Warning: {
      highlightColor = Color.Yellow;
      break;
    }
  }

  const codeSpan: Array<ScribblerJsx.Element> = [];

  for (let index = firstLine; index <= lastLine; index++) {
    const lineStart = diagnosticOrigin.sourceFile.getPositionOfLineAndCharacter(index, 0);
    const lineEnd =
      index === lastLineInFile
        ? diagnosticOrigin.sourceFile.text.length
        : diagnosticOrigin.sourceFile.getPositionOfLineAndCharacter(index + 1, 0);

    const lineText = diagnosticOrigin.sourceFile.text.slice(lineStart, lineEnd).trimEnd().replace(/\t/g, " ");

    if (index >= firstMarkedLine && index <= lastMarkedLine) {
      codeSpan.push(
        <CodeLineText
          gutterWidth={gutterWidth}
          lineNumber={index + 1}
          lineNumberColor={highlightColor}
          lineText={lineText}
        />,
      );

      if (index === firstMarkedLine) {
        const squiggleLength =
          index === lastMarkedLine
            ? lastMarkedLineCharacter - firstMarkedLineCharacter
            : lineText.length - firstMarkedLineCharacter;

        codeSpan.push(
          <SquiggleLineText
            gutterWidth={gutterWidth}
            indentWidth={firstMarkedLineCharacter}
            squiggleColor={highlightColor}
            squiggleWidth={squiggleLength}
          />,
        );
      } else if (index === lastMarkedLine) {
        codeSpan.push(
          <SquiggleLineText
            gutterWidth={gutterWidth}
            squiggleColor={highlightColor}
            squiggleWidth={lastMarkedLineCharacter}
          />,
        );
      } else {
        codeSpan.push(
          <SquiggleLineText gutterWidth={gutterWidth} squiggleColor={highlightColor} squiggleWidth={lineText.length} />,
        );
      }
    } else {
      codeSpan.push(<CodeLineText gutterWidth={gutterWidth} lineNumber={index + 1} lineText={lineText} />);
    }
  }

  const location = (
    <Line>
      {" ".repeat(gutterWidth + 2)}
      <Text color={Color.Gray}>{" at "}</Text>
      <Text color={Color.Cyan}>{Path.relative("", diagnosticOrigin.sourceFile.fileName)}</Text>
      <Text color={Color.Gray}>{`:${firstMarkedLine + 1}:${firstMarkedLineCharacter + 1}`}</Text>
      {diagnosticOrigin.assertion && <BreadcrumbsText ancestor={diagnosticOrigin.assertion.parent} />}
    </Line>
  );

  return (
    <Text>
      {codeSpan}
      <Line />
      {location}
    </Text>
  );
}
