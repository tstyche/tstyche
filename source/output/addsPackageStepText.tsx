import { Color, Line, Text } from "#scribbler";

export function addsPackageStepText(compilerVersion: string, installationPath: string): JSX.Element {
  return (
    <Line>
      <Text color={Color.Gray}>{"adds"}</Text>
      {" TypeScript "}
      {compilerVersion}
      <Text color={Color.Gray}>
        {" to "}
        {installationPath}
      </Text>
    </Line>
  );
}
