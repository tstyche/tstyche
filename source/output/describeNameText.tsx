import { Line } from "#scribbler";

export function describeNameText(name: string, indent = 0): JSX.Element {
  return <Line indent={indent + 1}>{name}</Line>;
}
