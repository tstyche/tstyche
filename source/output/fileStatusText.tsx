import type { TestFile } from "#file";
import { Path } from "#path";
import { type FileResultStatus, ResultStatus } from "#result";
import { Color, Line, type ScribblerJsx, Text } from "#scribbler";

interface FileNameTextProps {
  filePath: string;
}

function FileNameText({ filePath }: FileNameTextProps) {
  const relativePath = Path.relative("", filePath);
  const lastPathSeparator = relativePath.lastIndexOf("/");

  const directoryNameText = relativePath.slice(0, lastPathSeparator + 1);
  const fileNameText = relativePath.slice(lastPathSeparator + 1);

  return (
    <Text>
      <Text color={Color.Gray}>{directoryNameText}</Text>
      {fileNameText}
    </Text>
  );
}

export function fileStatusText(status: FileResultStatus, testFile: TestFile): ScribblerJsx.Element {
  let statusColor: Color;
  let statusText: string;

  switch (status) {
    case ResultStatus.Runs: {
      statusColor = Color.Yellow;
      statusText = "runs";
      break;
    }
    case ResultStatus.Passed: {
      statusColor = Color.Green;
      statusText = "pass";
      break;
    }
    case ResultStatus.Failed: {
      statusColor = Color.Red;
      statusText = "fail";
      break;
    }
  }

  return (
    <Line>
      <Text color={statusColor}>{statusText}</Text> <FileNameText filePath={testFile.path} />
    </Line>
  );
}
