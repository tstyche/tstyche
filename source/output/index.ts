export { addsPackageStepText } from "./addsPackageStepText.js";
export { describeNameText } from "./describeNameText.js";
export { diagnosticText } from "./diagnosticText.js";
export { fileStatusText } from "./fileStatusText.js";
export { fileViewText } from "./fileViewText.js";
export { formattedText } from "./formattedText.js";
export { helpText } from "./helpText.js";
export { OutputService, type OutputServiceOptions, type WriteStream } from "./OutputService.js";
export { summaryText } from "./summaryText.js";
export { testNameText } from "./testNameText.js";
export { usesCompilerStepText } from "./usesCompilerStepText.js";
export { watchUsageText } from "./watchUsageText.js";
