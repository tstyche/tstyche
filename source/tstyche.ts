export * from "#api";
export * from "#cli";
export * from "#collect";
export * from "#config";
export * from "#diagnostic";
export * from "#environment";
export * from "#events";
export * from "#expect";
export * from "#logger";
export * from "#output";
export * from "#path";
export * from "#project";
export * from "#reporters";
export * from "#result";
export * from "#runner";
export * from "#scribbler";
export * from "#select";
export * from "#store";
export * from "#token";
export * from "#version";
export * from "#watcher";
