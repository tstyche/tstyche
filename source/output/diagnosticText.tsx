import { type Diagnostic, DiagnosticCategory } from "#diagnostic";
import { Color, Line, type ScribblerJsx, Text } from "#scribbler";
import { CodeSpanText } from "./CodeSpanText.js";

interface DiagnosticTextProps {
  diagnostic: Diagnostic;
}

function DiagnosticText({ diagnostic }: DiagnosticTextProps) {
  const code = typeof diagnostic.code === "string" ? <Text color={Color.Gray}> {diagnostic.code}</Text> : undefined;

  const text = Array.isArray(diagnostic.text) ? diagnostic.text : [diagnostic.text];

  const message = text.map((text, index) => (
    <Text>
      {index === 1 ? <Line /> : undefined}
      <Line>
        {text}
        {code}
      </Line>
    </Text>
  ));

  const related = diagnostic.related?.map((relatedDiagnostic) => <DiagnosticText diagnostic={relatedDiagnostic} />);

  const codeSpan = diagnostic.origin ? (
    <Text>
      <Line />
      <CodeSpanText diagnosticOrigin={diagnostic.origin} />
    </Text>
  ) : undefined;

  return (
    <Text>
      {message}
      {codeSpan}
      <Line />
      <Text indent={2}>{related}</Text>
    </Text>
  );
}

export function diagnosticText(diagnostic: Diagnostic): ScribblerJsx.Element {
  let prefix: ScribblerJsx.Element | undefined;

  switch (diagnostic.category) {
    case DiagnosticCategory.Error: {
      prefix = <Text color={Color.Red}>{"Error: "}</Text>;
      break;
    }

    case DiagnosticCategory.Warning: {
      prefix = <Text color={Color.Yellow}>{"Warning: "}</Text>;
      break;
    }

    default:
      break;
  }

  return (
    <Text>
      {prefix}
      <DiagnosticText diagnostic={diagnostic} />
    </Text>
  );
}
