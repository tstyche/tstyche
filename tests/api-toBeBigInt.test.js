import { test } from "poku";
import * as tstyche from "tstyche";
import * as assert from "./__utilities__/assert.js";
import { getFixtureFileUrl, getTestFileName } from "./__utilities__/fixture.js";
import { normalizeOutput } from "./__utilities__/output.js";
import { spawnTyche } from "./__utilities__/tstyche.js";

const testFileName = getTestFileName(import.meta.url);
const fixtureUrl = getFixtureFileUrl(testFileName);

test("'toBeBigInt' implementation", () => {
  tstyche.expect(BigInt(123)).type.toBeBigInt();
  tstyche.expect(123).type.not.toBeBigInt();
});

await test("toBeBigInt", async () => {
  const { exitCode, stderr, stdout } = await spawnTyche(fixtureUrl);

  await assert.matchSnapshot(normalizeOutput(stdout), {
    fileName: `${testFileName}-stdout`,
    testFileUrl: import.meta.url,
  });

  await assert.matchSnapshot(stderr, {
    fileName: `${testFileName}-stderr`,
    testFileUrl: import.meta.url,
  });

  assert.equal(exitCode, 1);
});
