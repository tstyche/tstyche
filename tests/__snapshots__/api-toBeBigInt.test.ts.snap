// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toBeBigInt: stderr 1`] = `
"Error: Type 'bigint' is not identical to type 'void'.

  12 |   expect(returnsBigInt()).type.toBeBigInt();
  13 | 
> 14 |   expect(returnsVoid()).type.toBeBigInt();
     |                              ^
  15 | });
  16 | 
  17 | test("is NOT bigint?", () => {

       at ./__typetests__/toBeBigInt.test.ts:14:30 ❭ is bigint?

Error: Type 'bigint' is identical to type 'bigint'.

  18 |   expect(returnsVoid()).type.not.toBeBigInt();
  19 | 
> 20 |   expect(returnsBigInt()).type.not.toBeBigInt();
     |                                    ^
  21 | });
  22 | 

       at ./__typetests__/toBeBigInt.test.ts:20:36 ❭ is NOT bigint?

"
`;

exports[`toBeBigInt: stdout 1`] = `
"uses TypeScript <<version>> with ./tsconfig.json

fail ./__typetests__/toBeBigInt.test.ts
  × is bigint?
  × is NOT bigint?

Targets:    1 failed, 1 total
Test files: 1 failed, 1 total
Tests:      2 failed, 2 total
Assertions: 2 failed, 2 passed, 4 total
Duration:   <<timestamp>>

Ran all test files.
"
`;
